cpu/esp32/esp-now/esp_now_netdev.c

```diff
@@ -0,0 +1,872 @@
+    _esp_now_scan_peers_timer.arg = dev;
+
+    /* execute the first scan */
+    esp_now_scan_peers_done();
+
+    #else /* ESP_NOW_UNICAST */
+    #if 0
+    int res = esp_now_add_peer((uint8_t*)_esp_now_mac, ESP_NOW_ROLE_COMBO,
+                               esp_now_params.channel, NULL, 0);
+    DEBUG("%s: multicast node added %d\n", __func__, res);
+    #endif
+    #endif /* ESP_NOW_UNICAST */
+}
+
+static int _init(netdev_t *netdev)
+{
+    DEBUG("%s: %p\n", __func__, netdev);
+
+    #ifdef MODULE_NETSTATS_L2
+    memset(&netdev->stats, 0x00, sizeof(netstats_t));
+    #endif
+
+    return 0;
+}
+
+static int _send(netdev_t *netdev, const iolist_t *iolist)
+{
+    #if ESP_NOW_UNICAST
+    if (!_esp_now_scan_peers_done) {
+        return -ENODEV;
+    }
+    #endif
+
+    DEBUG("%s: %p %p\n", __func__, netdev, iolist);
+
+    CHECK_PARAM_RET (netdev != NULL, -ENODEV);
+    CHECK_PARAM_RET (iolist != NULL, -EINVAL);
+
+    esp_now_netdev_t* dev = (esp_now_netdev_t*)netdev;
+
+    mutex_lock(&dev->dev_lock);
+    dev->tx_len = 0;
+
+    /* load packet data into TX buffer */
+    for (const iolist_t *iol = iolist; iol; iol = iol->iol_next) {
+        if (dev->tx_len + iol->iol_len > ESP_NOW_MAX_SIZE) {
+            mutex_unlock(&dev->dev_lock);
+            return -EOVERFLOW;
+        }
+        memcpy (dev->tx_buf + dev->tx_len, iol->iol_base, iol->iol_len);
+        dev->tx_len += iol->iol_len;
+    }
+
+    #if ENABLE_DEBUG
+    printf ("%s: send %d byte\n", __func__, dev->tx_len);
+    /* esp_hexdump (dev->tx_buf, dev->tx_len, 'b', 16); */
+    #endif
+
+    _esp_now_sending = 1;
+
+    uint8_t* _esp_now_dst = 0;
+
+    #if ESP_NOW_UNICAST
+    ipv6_hdr_t* ipv6_hdr = (ipv6_hdr_t*)dev->tx_buf;
+    uint8_t  _esp_now_dst_from_iid[6];
+
+    if (ipv6_hdr->dst.u8[0] == 0xff) {
+        /* packets to multicast prefix ff::/8 are sent to all peers */
+        DEBUG("multicast to all peers\n");
+        _esp_now_dst = 0;
+        _esp_now_sending = dev->peers_all;
+
+        #ifdef MODULE_NETSTATS_L2
+        netdev->stats.tx_mcast_count++;
+        #endif
+    }
+
+    else if ((byteorder_ntohs(ipv6_hdr->dst.u16[0]) & 0xffc0) == 0xfe80) {
+        /* for link local addresses fe80::/10, the MAC address is derived from dst address */
+        _get_mac_from_iid(&ipv6_hdr->dst.u8[8], _esp_now_dst_from_iid);
+        DEBUG("link local to %02x:%02x:%02x:%02x:%02x:%02x\n",
+              _esp_now_dst_from_iid[0], _esp_now_dst_from_iid[1],
+              _esp_now_dst_from_iid[2], _esp_now_dst_from_iid[3],
+              _esp_now_dst_from_iid[4], _esp_now_dst_from_iid[5]);
+        _esp_now_dst = _esp_now_dst_from_iid;
+        _esp_now_sending = 1;
+    }
+
+    else {
+        #ifdef MODULE_GNRC_IPV6_NIB
+        /* for other addresses, try to find an entry in NIB cache */
+        gnrc_ipv6_nib_nc_t nce;
+        int ret = gnrc_ipv6_nib_get_next_hop_l2addr (&ipv6_hdr->dst, dev->netif,
```

Comment:
What is IPv6 stuff doing here oO

Reply:
Good question, it is long time ago :worried:

As I remember, the first packet snippet the `netdev` driver gets in the `_send` function is the IPv6 header. If the IPv6 destination address is not a link local address, the destination MAC address cannot be derived from the IPv6 destination address. Of course, I would have the same problem when privacy extensions are used. So I thought, that `nib` is the right way to get the mapping from an IPv6 destination address to the next hop MAC address. Maybe, I was wrong. In that case the question is how it should be.

From my understanding, IPv6 is responsible to find the next hop IPv6 address from the routing table and to use the neighbor discovery protocol to determine the next hop MAC address. This MAC address is then given as destination parameter of the MA_UNIDATA.request primitive to the L2 interface.

I couldn't figure out how this is exactly realized in GNRC.

Reply:
This is not supposed to go in the network device. GNRC's IPv6 module should take care of all of that. If the link-layer is not multicast/broadcast capable, special care needs to be taken anyway.

Reply:
This is what I feared. ESP-NOW is unfortunatly a peer-to-peer network and not a meshed network :-( That is, each message would have to be sent to each known peer as MAC unicast to simulate the broadcast. I thought I could do this hack with nib to save time by trying to figure out the peer and then only to send the message as a single unicast message.

Reply:
@miri64 
You said that the whole file looks like a hack. That might be, but the ESP32 port was a big project in parallel to my regular business and I was happy to get it working with WiFi at all even if it is not yet perfect. Now, where I will have some time over the comming holiday season, I would like to improve or correct it.

Therefore, I would like to ask you to give me a short guide which modules have to be used and how to get a MAC header as the first packet snippet in the `_send` function of the netdev driver, especially if I have Ethernet addresses, but none MAC according to IEEE 805.14 or IEEE 802.3.

I tried again to figure out how GNRC IPv6 works with `netdev_eth` and `enc28j60`, but I'm still not sure. Do I understand it correctly that the `_send` function of `enc28j60` receives a preformatted IEEE802.3 or Ethernet II MAC header as the first packet snippet in the parameter `iolist`? If so, would it be an appropriate way to change the `esp_now netdev` driver to be rather an adoption layer that emulates an Ethernet MAC layer over ESP-NOW and simply to use it like an `netdev_eth` device?

Reply:
Ok, I probably did not express myself precisely enough. The question was of course, whether I'm right that an Ethernet driver can rely upon that the flattened `iolist` begins with the 14 byte Ethernet MAC header.

I just need some guidance on what the right approach is to realize a netdev driver for a network technology that is not IEEE802.15.4 or IEEE802.3 conformant to use it with GNRC.

Reply:
> Ok, I probably did not express myself precisely enough. The question was of course, whether I'm right that an Ethernet driver can rely upon that the flattened `iolist` begins with the 14 byte Ethernet MAC header.

Yes!

> I just need some guidance on what the right approach is to realize a netdev driver for a network technology that is not IEEE802.15.4 or IEEE802.3 conformant to use it with GNRC.

IMO using ethernet as abstraction might not be the best choice, as on ethernet we usually use full-blown IPv6 and large MTUs.

Maybe take a look at how cc1100, nrfmin and the coming nimble gnrc integrations are handling non-standard link layers?

The netdev _send function should expect whatever the underlying device needs. A small espnow_gnrc layer would handle the conversion of the generic netif header to that iolist. See e.g., ```cpu/nrf5x_common/radio/nrfmin/nrfmin_gnrc.c``` as an example.

How does the ESP-NOW send function look like? If it also requires data blocks (or snippets) in a specific format, obviously re-use that. 
If it has a signature like "esp_now_send(src_addr, target_addr, other_stuff, data, data_len)" (with "header" parts needed as parameters), it might make sense to define a struct with the necessary fields and expect that struct at the beginning of what is passed to netdev::send().

Reply:
@kaspar030 Many thanks for your hints. @TimoRoth also suggested in issue #10531  to realize the ESP-NOW netdev in a similar way as the cc110x netdev.

> How does the ESP-NOW send function look like?


Since ESP-NOW is a peer-to-peer network, ESP_NOW netdev driver has unfortunatly do much more than simply to send and receive. It has also to manage the connections to visible peers.

Reply:
I got an initial (yet fully untested beyond compiling it) version of the rework of this on my fork:
https://github.com/TimoRoth/RIOT

Reply:
I got an initial (but yet fully untested beyond compiling it) version of the rework of this on https://github.com/TimoRoth/RIOT

Reply:
> Am I right that the reason why ESP-NOW gets the IPv6 header as the first header instead of the netif header with the required L2 information because it creates a netif thread of type raw with function gnrc_netif_raw_create?

Yepp. (also as a side-note: the current `gnrc_netif` concept is a bit different and younger than the state of my master thesis)

Reply:
But to restate my comment from https://github.com/RIOT-OS/RIOT/issues/10531#issuecomment-445805488:

> The `gnrc_netif_raw` implementation was intended for link-layers without any addresses (e.g. SLIP) or were access to the link-layer address is not necessary.

Reply:
> I got an initial (but yet fully untested beyond compiling it) version of the rework of this on https://github.com/TimoRoth/RIOT

Now provided as PR to discuss on https://github.com/RIOT-OS/RIOT/pull/10581.